### PR TITLE
close command implementation

### PR DESCRIPTION
**Pull request by baldarn**

### CODE REVIEW

File: src/connection.ts [hunk 1]
```diff
@@ -75,10 +77,15 @@ export class Connection {
...
-  public close(): Promise<void> {
+  public async close(
+    params: { closingCode: number; closingReason: string } = { closingCode: 0, closingReason: "" }
+  ): Promise<void> {
     this.logger.info(`Closing connection ...`)
     this.heartbeat.stop()
-    return new Promise((res, _rej) => this.socket.end(() => res()))
+    this.logger.debug(`Close ...`)
+    const res = await this.sendAndWait<CloseResponse>(new CloseRequest(params))
+    this.socket.end()
```
Comment (gpad) [inline src/connection.ts:87]: You have to return when the `end` is completed, see the old code:

```typescript
    return new Promise((res, _rej) => this.socket.end(() => res()))
```

File: test/e2e/basic_publish.test.ts [hunk 2]
```diff
@@ -32,7 +32,6 @@ describe("publish a message", () => {
     await eventually(async () => {
       expect((await rabbit.getQueueInfo(stream)).messages).eql(1)
     }, 10000)
-    await connection.close()
```
Comment (gpad) [inline test/e2e/basic_publish.test.ts:35]: Why do you remove this code?

File: test/e2e/basic_publish.test.ts [hunk 3]
```diff
@@ -47,6 +46,5 @@ describe("publish a message", () => {
     await eventually(async () => {
       expect((await rabbit.getQueueInfo(stream)).messages).eql(100)
     }, 10000)
-    await connection.close()
```
Comment (gpad) [inline test/e2e/basic_publish.test.ts:50]: Why do you remove this code?

File: test/e2e/basic_publish.test.ts (hunk 2 above)
Comment (baldarn) [reply test/e2e/basic_publish.test.ts:35]: the
```
afterEach(() => connection.close())
```

closes the connection 2 times, giving an error now that the close on the rabbit level is done. do you think we have to handle that differently?

File: test/e2e/basic_publish.test.ts (hunk 3 above)
Comment (baldarn) [reply test/e2e/basic_publish.test.ts:50]: the
```
afterEach(() => connection.close())
```

closes the connection 2 times, giving an error now that the close on the rabbit level is done. do you think we have to handle that differently?

File: test/e2e/basic_publish.test.ts (hunk 2 above)
Comment (gpad) [reply test/e2e/basic_publish.test.ts:35]: mmh you are right I think it's fine in this way

File: test/e2e/basic_publish.test.ts (hunk 3 above)
Comment (gpad) [reply test/e2e/basic_publish.test.ts:50]: as above

**Review by gpad — APPROVED** [review]

**Pull request merged**